### PR TITLE
feat: Customized functions in runtime should pass validation and be allowed to start bot

### LINKED ISSUE #3656: Customized functions in runtime should pass validation and be allowed to start bot
## Is your feature request related to a problem? Please describe.

Currently, all function calls in an expression is parsed and checked to make sure it's a valid function, otherwise we will throw error and block user from starting bot. 

The issue is for some customers, if they have a customized runtime that have customized functions (using code) built-in and composer isn't aware of that, thus composer will still treat it as an error. 

## Describe the solution you'd like

Add a settings field "customizedFunctions" (or any other name) to allow user to config a list of functions name that is customized and should be ignored in validation phase.

### PR DESCRIPTION
**Pull request by lei9444**
## Description
Add a settings field "customizedFunctions" to allow user to config a list of functions name that is customized and should be ignored in validation phase.

## Task Item
closes #3656
refs #3130

## Screenshots
custom

### CODE REVIEW

**Review by cwhitten — DISMISSED** [review]

**Comment by boydc2014** [comment]
Current in master, I think we ignore the error of undefined functions, right? So some customer may not aware of they are using a customized functions. So can we at least make sure we don't break MeetAssitant, maybe give them a list of functions they need to add?

**Comment by cwhitten** [comment]
@lei9444 can you address conflicts please

File: Composer/packages/lib/indexers/src/validations/expressionValidation/validation.ts [hunk 1]
```diff
@@ -14,9 +14,12 @@ const RETURNTYPE_NOT_MATCH = formatMessage('the return type does not match');
 const BUILT_IN_FUNCTION_ERROR = formatMessage("it's not a built-in function or a custom function.");
 
 const expressionErrorMessage = (error: string) => formatMessage('must be an expression: {error}', { error });
-
+const builtInFunctionErrorMessage = (error: string) =>
+  formatMessage(`{error} Please add unknow functions to setting's customFunctions field.`, {
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/indexers/src/validations/expressionValidation/validation.ts:18]:
```suggestion
  formatMessage(`{error} Please add unknown functions to setting's customFunctions field.`, {
```

File: Composer/packages/lib/indexers/src/validations/index.ts [hunk 2]
```diff
@@ -13,68 +12,59 @@ export const validateFuncs: { [type: string]: ValidateFunc[] } = {
...
+        const allChecks = [...validateFuncs['.']];
+        const checkerFunc = validateFuncs[value.$kind];
+        if (checkerFunc) {
+          allChecks.splice(0, 0, ...checkerFunc);
+        }
+
+        allChecks.forEach((func) => {
+          const result = func(path, value, value.$kind, schema.definitions[value.$kind], settings, lgFiles, luFiles);
+          if (result) {
+            diagnostics.splice(0, 0, ...result);
```
Comment (a-b-r-o-w-n) [inline Composer/packages/lib/indexers/src/validations/index.ts:42]: Is there a reason we slice the result into diagnostics? Seems to warrant a comment if so and if not, this should just be a `push` instead.

**Comment by cwhitten** [comment]
ping @lei9444

**Comment by cwhitten** [comment]
FYI @zxyanliu. let's add docs work item to explain this new feature

**Comment by cwhitten** [comment]
overriding the E2E test failure. It looks like a cache issue with the devops agent.

**Review by cwhitten — APPROVED** [review]

**Comment by zxyanliu** [comment]
Got it. I will follow up with @cwhitten and @lei9444 on this soon.

**Comment by zxyanliu** [comment]
@cwhitten and @lei9444 
The content has been added and [the PR](https://github.com/MicrosoftDocs/composer-docs-pr/pull/288) is ready for review.

**Comment by zxyanliu** [comment]
FYI. The doc on this feature has been merged to master. The content will go live soon.

**Comment by TheDreamSaver** [comment]
Hi @zxyanliu
Is the doc on this live yet?

**Pull request merged**